saged/models.py

```diff
@@ -811,3 +862,180 @@ def save_model(self, out_path: str) -> None:
+        experiment_name = self.experiment_name
+        experiment_description = self.experiment_description
+        log_progress = self.log_progress
+
+        train_count = None
+        train_fraction = None
+        train_fraction = getattr(self, 'train_fraction', None)
+        if train_fraction is None:
+            train_count = self.train_count
+
+        labeled_data = dataset.get_labeled()
+        unlabeled_data = dataset.get_unlabeled()
+
+        # Split dataset and create dataloaders
+        train_dataset, tune_dataset = labeled_data.train_test_split(train_fraction=train_fraction,
+                                                                    train_study_count=train_count,
+                                                                    seed=seed)
+        train_loader = DataLoader(train_dataset, batch_size, shuffle=True, drop_last=True)
+        tune_loader = DataLoader(tune_dataset, batch_size=1)
+        unlabeled_loader = DataLoader(unlabeled_data, batch_size, shuffle=True, drop_last=True)
+
+        self.model.to(device)
+
+        self.loss_fn = self.loss_class()
+        # If the loss function is weighted, weight losses based on the classes' prevalance
+
+        if torch.nn.modules.loss._WeightedLoss in self.loss_class.__bases__:
+            # TODO calculate class weights
+            self.loss_fn = self.loss_class(weight=None)
+
+        if log_progress:
+            experiment = neptune.create_experiment(name=experiment_name,
+                                                   description=experiment_description,
+                                                   params=self.config
+                                                   )
+
+            experiment.set_property('model', str(type(self.model)))
+
+            # Track the baseline (always predicting the most common class)
+            label_counts = tune_dataset.map_labels_to_counts().values()
+
+            tune_baseline = max(label_counts) / sum(label_counts)
+            neptune.log_metric('tune_baseline', tune_baseline)
+
+        best_tune_loss = None
+
+        for epoch in tqdm(range(epochs)):
+
+            # Alpha (the ratio of pseudolabel loss to label loss to use
+            # increases linearly across epochs. This allows the model to ignore pseudolabeling
+            # information while the model is bad, and increases pseudolabels' impact when
+            # the model has been trained on the input data for awhile
+            if epochs > 1:
+                progress = epoch / (epochs-1)
+            else:
+                progress = 0
+            alpha = progress * self.max_alpha
+
+            train_loss = 0
+            train_correct = 0
+            self.model.train()
+
+            # Create an iterator that returns a labeled batch and an unlabeled batch
+            if len(labeled_data) <= len(unlabeled_data):
+                # If there are more unlabeled samples than labeled samples, just stop iterating
```

Comment:
Also, I'm a bit confused by the comment here - isn't each of these a `DataLoader`? So wouldn't you be concerned about the number of batches in each, not the number of samples?

Probably not a big deal since you seem to be using the same batch size for the supervised and unsupervised loaders, but it might be good to make a note of this in case you vary batch sizes in the future.

Reply:
That's good to check, but `labeled_data` and `unlabeled_data` are the datasets with `train_loader` and `unlabeled_loader` are their corresponding loaders. They are created on lines 914 and 915 from the dataset passed in